### PR TITLE
fix(behavior_path_planner): add distance to returned point in avoidance module

### PR DESCRIPTION
**Pull request by shulanbushangshu**
## Description

To solve #833
We can calculate the maximum longitudinal distance in the direction of the lane according to the part of the obstacle in the lane. Append the calculated maximum longitudinal distance to the longitudinal starting distance of the return point.
The schematic of the change is shown below：
2022-05-30 11-27-12屏幕截图

## Related links

#833
## Tests performed

1.Run planning_simulator.
2.place ego position and goal
3. Set a obstacle near lanes
4. result -- the path will maintain the calculated offset for the horizontal distance where obstacle is present
2022-05-13 13-58-50屏幕截图

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Comment by shulanbushangshu** [comment]
@zulfaqar-azmi-t4 @TakaHoribe ,can you review this pr ? I add distance by considering the length of obstacle and the lane to the returned point in the avoidance module of behavior_path_planner.

**Comment by taikitanaka3** [comment]
@shulanbushangshu 
Can you add the desctiption of this change with figure at PR description or issue, and please update avoidace-design.md according to agreed chage?

File: planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp [hunk 1]
```diff
@@ -2676,5 +2679,86 @@ void AvoidanceModule::setDebugData(const PathShifter & shifter, const DebugData
...
+  }
+
+  check_points.insert(check_points.end(), collision_points.begin(), collision_points.end());
+  check_points.insert(check_points.end(), in_lane_points.begin(), in_lane_points.end());
 
+  for (auto & check_point : check_points) {
+    const auto dx = check_point.x() - object_pose.position.x;
+    const auto dy = check_point.y() - object_pose.position.y;
+    const Eigen::Vector3d diff_vec{dx, dy, 0};
+    const Eigen::Vector3d cross_vec = base_unit_vec.cross(diff_vec);
```
Comment (taikitanaka3) [inline planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp:2749]: @shulanbushangshu 
Thank you for description can you also add how long is "distance_add_to_returnpoint" ?

and I suggest these changes (please give me opinion if I'm wrong.)
- I think this name( "distance_add_to_returnpoint") should be something like longitudinal_collision_margin
- longitudinal margin should be added to NOT closer footprint point because it's a bug which should be fixed.
image

Comment (shulanbushangshu) [reply planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp:2749]: @taikitanaka3
Yes,you are right.In the changes,the "distance_add_to_returnpoint" is calculated according to the check points (including the intersections of the obstacle and the lane boundary, and the contour point of the obstacle within the lane).
The "distance_add_to_returnpoint" can be as longitudinal_collision_margin.
The schematic of the  "distance_add_to_returnpoint" is shown below：
distance

In the code,I use the check points to calculate the maximum distance in the "object_closest_pose" point direction.The maximum distance is the "distance_add_to_returnpoint"

**Comment by zulfaqar-azmi-t4** [comment]
Test is conducted based on approval by default. 
It seems no issue for now.

Screenshot from 2022-05-31 23-06-34

**Comment by zulfaqar-azmi-t4** [comment]
Since the length of the bus is a little longer, i think we should also consider the following case and properly define the suitable behavior. Referring to your schematics, I think for this case, point B and C should be adjust accordingly.  

https://user-images.githubusercontent.com/93502286/171196007-2d17f4e0-c3bc-454f-a8f0-2dd60c29d840.mp4

File: planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp [hunk 2]
```diff
@@ -573,9 +576,9 @@ AvoidPointArray AvoidanceModule::calcRawShiftPointsFromObjects(
     AvoidPoint ap_return;
     ap_return.length = 0.0;
     ap_return.start_length = shift_length;
-    ap_return.start_longitudinal = o.longitudinal;
-    ap_return.end_longitudinal =
-      o.longitudinal + std::min(nominal_return_distance, return_remaining_distance);
+    ap_return.start_longitudinal = o.longitudinal + o.distance_add_to_returnpoint;
```
Comment (taikitanaka3) [inline planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp:579]: @shulanbushangshu 
why do you add longitudinal distance to start _longitudinal?

File: planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp (hunk 1 above)
Comment (taikitanaka3) [reply planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp:2749]: @shulanbushangshu 
can you add better name for check point or add description to readme?

@zulfaqar-azmi-t4 
can you share figures with @shulanbushangshu used in readme?

Comment (taikitanaka3) [reply planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp:2749]: @shulanbushangshu 
can you add better name for check point or add description to readme?

@zulfaqar-azmi-t4 
can you share figures with @shulanbushangshu used in readme?

File: planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (shulanbushangshu) [reply planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp:579]: @taikitanaka3 
I want to separate the longitudinal end of the starting point from the longitudinal start of the return point so that "BC" can maintain the maximum offset.
2022-06-02 09-59-42屏幕截图.

[the remark this answers is not in the document]
Comment (zulfaqar-azmi-t4) [reply planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp:579]: @shulanbushangshu 
My apologies, I might misunderstood your intention. Assuming I take this [image](https://user-images.githubusercontent.com/102840938/171134680-95943418-72cc-4e34-9374-bf2d530cdf79.png) as the example, it essentially mean you're shifting the start_longitudinal pose from point B to point C. 

Is it possible if you can provide some sketch (simple is also fine) regarding maintaining the maximum offset for "BC so that we can better understand the situation?

File: planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp (hunk 1 above)
Comment (zulfaqar-azmi-t4) [reply planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp:2749]: @taikitanaka3 
Yes, this is the drawio file.

[feature_behavior_path_planner.zip](https://github.com/autowarefoundation/autoware.universe/files/8821650/feature_behavior_path_planner.zip)

File: planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (shulanbushangshu) [reply planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp:579]: @zulfaqar-azmi-t4 @taikitanaka3 
Thank you for your review.Let me briefly describe the process of issue#833:
In the process of obstacle avoidance simulation, I found that the default lateral safety margin was a little large. When the intersection between obstacle and lane was very small, there would be an offset of about 1.5m, as shown in the figure below.
large

Therefore, I changed the default lateral safety margin and conducted simulation on this basis.There are two variables :''lateral_collision_margin" and "lateral_collision_safety_buffer".I set "lateral_collision_safety_buffer" to 0.0.
In the following simulation, issue#833 occurs, as shown below:
2022-06-02 15-21-46屏幕截图

Of course, if the lateral safety distance margin is enlarged, there may be no collision. But I think it is possible to reduce the risk of parameter adjustment by calculating the compensation amount of a longitudinal distance.According to the above situation, I calculated the compensation amount of the longitudinal distance according to the obstacles and lanes.I want to separate the longitudinal end of the starting point from the longitudinal start of the return point so that "BC" can maintain the maximum offset.
The distance is calculated as follows:
2022-06-02 15-47-13屏幕截图
In this example,the 'distance_add_to_returnpoint'='d2'+'d4'.
Calculation principle: calculate the minimum distance and maximum distance from the contour point and intersection point to the reference point. The maximum distance minus the minimum distance is used as the compensation distance.
Please consider whether my idea is necessary and correct.

[the remark this answers is not in the document]
Comment (zulfaqar-azmi-t4) [reply planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp:579]: @shulanbushangshu 
About this [image](https://user-images.githubusercontent.com/102840938/171581642-ccf2b61f-3e75-40d4-a498-50c7ffebc64b.png), the result indeed seems weird. it is possible for you to reproduce them with the debug marker `publish_debug_marker=true` in the `avoidance.param.yaml` and share the image with us?

[the remark this answers is not in the document]
Comment (taikitanaka3) [reply planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp:579]: @shulanbushangshu 
I've tested your code but it looks like something wrong
image (6)

[the remark this answers is not in the document]
Comment (taikitanaka3) [reply planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp:579]: In your algorithm I think you need to add margin for start longitudinal too according to collision of foot print and lanelet.

[the remark this answers is not in the document]
Comment (shulanbushangshu) [reply planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp:579]: @zulfaqar-azmi-t4 @taikitanaka3 
Thank you for reviewing.
This is the debug marker:
2022-06-06 14-30-58屏幕截图
@taikitanaka3 ,you are right.I forgot to change the vertical distance of the starting point.I will fix it.
The two issues((#833,#934)) I raised are aimed at different problems.
The first is to solve the problem of early return (there will be problems if there are long obstacles).
The second is to solve the offset distance (there may be problems in calculating the offset distance in some cases) We also consider the points on the contour of the obstacle and the distance to the boundary. I think it is necessary to modify the calculation method of lateral offset distance.
To sum up, the first issue can be solved by longitudinal distance compensation; The second issue can be solved by lateral distance compensation.
Of course, the above two problems will lead to the risk of collision. In this PR, only the first issue is solved.

[the remark this answers is not in the document]
Comment (taikitanaka3) [reply planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp:579]: @shulanbushangshu 
Good, but at first I want to merge fix PR for considering vehicle size in frenet-coordinate is it ok?
your PR looks like new feature that we haven't tested yet.(for the reason you change first avoid point and second return point according to the collision of lanelet and collision edge case is not fully defined or tested yet)  
https://github.com/autowarefoundation/autoware.universe/pull/1041#issuecomment-1147039130

[the remark this answers is not in the document]
Comment (shulanbushangshu) [reply planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp:579]: @taikitanaka3.No problem
